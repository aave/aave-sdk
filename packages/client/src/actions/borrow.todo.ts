import { describe, expect, it } from 'vitest';

// THIS IS A VERY WIP

describe('Given an Aave Market', () => {
  describe('And a user with a supply position', () => {
    describe('When the user enables e-mode for that market', () => {
      it('Then it should be possible to borrow from the reserve', () => {
        expect(true).toBe(true);
      });
    });
  });
});
