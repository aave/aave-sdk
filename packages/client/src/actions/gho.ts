import {
  type ExecutionPlan,
  SavingsGhoBalanceQuery,
  type SavingsGhoBalanceRequest,
  SavingsGhoDepositQuery,
  type SavingsGhoDepositRequest,
  SavingsGhoWithdrawQuery,
  type SavingsGhoWithdrawRequest,
  type TokenAmount,
  type TransactionRequest,
} from '@aave/graphql';
import type { ResultAsync } from '@aave/types';
import type { AaveClient } from '../client';
import type { UnexpectedError } from '../errors';

/**
 * Fetches the current sGHO balance for a user.
 *
 * ```ts
 * const result = await savingsGhoBalance(client, {
 *   user: evmAddress('0x742d35cc…'),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The sGHO balance request parameters.
 * @returns The user's sGHO balance.
 */
export function savingsGhoBalance(
  client: AaveClient,
  request: SavingsGhoBalanceRequest,
): ResultAsync<TokenAmount, UnexpectedError> {
  return client.query(SavingsGhoBalanceQuery, { request });
}

/**
 * Creates a transaction to withdraw sGHO.
 *
 * ```ts
 * const result = await savingsGhoWithdraw(client, {
 *   amount: {
 *     exact: '1000',
 *   },
 *   sharesOwner: evmAddress('0x9abc…'),
 * }).andThen(sendWith(wallet))
 *
 * if (result.isErr()) {
 *   // Handle error, e.g. signing error, etc.
 *   return;
 * }
 *
 * // result.value: TxHash
 * ```
 *
 * @param client - Aave client.
 * @param request - The sGHO withdraw request parameters.
 * @returns The transaction request data to withdraw sGHO.
 */
export function savingsGhoWithdraw(
  client: AaveClient,
  request: SavingsGhoWithdrawRequest,
): ResultAsync<TransactionRequest, UnexpectedError> {
  return client.query(SavingsGhoWithdrawQuery, { request });
}

/**
 * Creates a transaction to deposit GHO into sGHO.
 *
 * ```ts
 * const result = await savingsGhoDeposit(client, {
 *   amount: {
 *     value: '1000',
 *   },
 *   depositor: evmAddress('0x9abc…'),
 * }).andThen(sendWith(wallet))
 *
 * if (result.isErr()) {
 *   // Handle error, e.g. insufficient balance, signing error, etc.
 *   return;
 * }
 *
 * // result.value: TxHash
 * ```
 *
 * @param client - Aave client.
 * @param request - The sGHO deposit request parameters.
 * @returns The transaction data, approval requirements, or insufficient balance error.
 */
export function savingsGhoDeposit(
  client: AaveClient,
  request: SavingsGhoDepositRequest,
): ResultAsync<ExecutionPlan, UnexpectedError> {
  return client.query(SavingsGhoDepositQuery, { request });
}
