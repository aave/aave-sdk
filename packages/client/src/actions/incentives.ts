import {
  MeritClaimRewardsQuery,
  type MeritClaimRewardsRequest,
} from '@aave/graphql';
import type { AaveClient } from '../client';

/**
 * Fetches Merit claim rewards for a user with the transaction request to claim them.
 *
 * ```ts
 * const result = await meritClaimRewards(client, {
 *   user: evmAddress('0x742d35ccâ€¦'),
 *   chainId: chainId(1),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The merit claim rewards request parameters.
 * @returns The transaction to claim merit rewards.
 */
export function meritClaimRewards(
  client: AaveClient,
  request: MeritClaimRewardsRequest,
) {
  return client.query(MeritClaimRewardsQuery, { request });
}
