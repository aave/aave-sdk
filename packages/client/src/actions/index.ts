export * from './markets';
export * from './misc';
export * from './permits';
export * from './reserve';
export * from './transactions';
export * from './user';
export * from './vaults';
