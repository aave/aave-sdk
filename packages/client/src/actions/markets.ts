import type { UnexpectedError } from '@aave/core';
import {
  type Market,
  MarketQuery,
  type MarketReservesRequestOrderBy,
  MarketsQuery,
  type MarketUserState,
  OrderDirection,
  UserMarketStateQuery,
  type UserMarketStateRequest,
} from '@aave/graphql';
import type { ChainId, EvmAddress, ResultAsync } from '@aave/types';
import type { AaveClient } from '../AaveClient';

export const defaultMarketReservesRequestOrderBy: MarketReservesRequestOrderBy =
  {
    tokenName: OrderDirection.Asc,
  };

export type MarketsRequest = {
  /**
   * The markets you want to see based on the chain ids.
   */
  chainIds: ChainId[];
  /**
   * The user viewing the market (e.g., the connected wallet).
   *
   * If not provided, user fields will not be included.
   */
  user?: EvmAddress;
  /**
   * The order by clause for the borrow reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  borrowsOrderBy?: MarketReservesRequestOrderBy;
  /**
   * The order by clause for the supply reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  suppliesOrderBy?: MarketReservesRequestOrderBy;
};

/**
 * Fetches all markets for the specified chain IDs.
 *
 * ```ts
 * const result = await markets(client, {
 *   chainIds: [chainId(1), chainId(8453)],
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The markets request parameters.
 * @returns The list of markets.
 */
export function markets(
  client: AaveClient,
  {
    chainIds,
    borrowsOrderBy = defaultMarketReservesRequestOrderBy,
    suppliesOrderBy = defaultMarketReservesRequestOrderBy,
    user,
  }: MarketsRequest,
): ResultAsync<Market[], UnexpectedError> {
  return client.query(MarketsQuery, {
    request: { chainIds, user },
    borrowsOrderBy,
    suppliesOrderBy,
  });
}

export type MarketRequest = {
  /**
   * The pool address for the market.
   */
  address: EvmAddress;

  /**
   * The chain id the market pool address is deployed on.
   */
  chainId: ChainId;
  /**
   * The user viewing the market (e.g., the connected wallet).
   *
   * If not provided, user fields will not be included.
   */
  user?: EvmAddress;
  /**
   * The order by clause for the borrow reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  borrowsOrderBy?: MarketReservesRequestOrderBy;
  /**
   * The order by clause for the supply reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  suppliesOrderBy?: MarketReservesRequestOrderBy;
};

/**
 * Fetches a specific market by address and chain ID.
 *
 * ```ts
 * const result = await market(client, {
 *   address: evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2'),
 *   chainId: chainId(1),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The market request parameters.
 * @returns The market data, or null if not found.
 */
export function market(
  client: AaveClient,
  {
    address,
    chainId,
    user,
    borrowsOrderBy = defaultMarketReservesRequestOrderBy,
    suppliesOrderBy = defaultMarketReservesRequestOrderBy,
  }: MarketRequest,
): ResultAsync<Market | null, UnexpectedError> {
  return client.query(MarketQuery, {
    request: { address, chainId, user },
    borrowsOrderBy,
    suppliesOrderBy,
  });
}

/**
 * Fetches user account market data across all reserves.
 *
 * ```ts
 * const result = await userMarketState(client, {
 *   market: evmAddress('0x1234…'),
 *   user: evmAddress('0x5678…'),
 *   chainId: chainId(1)
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The user market state request parameters.
 * @returns The user's market state.
 */
export function userMarketState(
  client: AaveClient,
  request: UserMarketStateRequest,
): ResultAsync<MarketUserState, UnexpectedError> {
  return client.query(UserMarketStateQuery, { request });
}
