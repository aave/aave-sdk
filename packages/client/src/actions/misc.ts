import {
  type Chain,
  ChainsQuery,
  HasProcessedKnownTransactionQuery,
  type HasProcessedKnownTransactionRequest,
  HealthQuery,
  type UsdExchangeRate,
  UsdExchangeRatesQuery,
  type UsdExchangeRatesRequest,
} from '@aave/graphql';
import type { ResultAsync } from '@aave/types';
import type { AaveClient } from '../client';
import type { UnexpectedError } from '../errors';

/**
 * Health check query.
 *
 * ```ts
 * const result = await health(client);
 * ```
 *
 * @param client - Aave client.
 * @returns True or false
 */
export function health(
  client: AaveClient,
): ResultAsync<boolean, UnexpectedError> {
  return client.query(HealthQuery, {});
}

/**
 * Fetches the list of supported chains.
 *
 * @param client - Aave client.
 * @returns The list of supported chains.
 */
export function chains(
  client: AaveClient,
): ResultAsync<Chain[], UnexpectedError> {
  return client.query(ChainsQuery, {});
}

/**
 * Fetches USD exchange rates for different tokens on a given market.
 *
 * ```ts
 * const result = await usdExchangeRates(client, {
 *   market: evmAddress('0x1234…'),
 *   underlyingTokens: [evmAddress('0x5678…'), evmAddress('0x90ab…')],
 *   chainId: chainId(1),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The USD exchange rates request parameters.
 * @returns The list of USD exchange rates.
 */
export function usdExchangeRates(
  client: AaveClient,
  request: UsdExchangeRatesRequest,
): ResultAsync<UsdExchangeRate[], UnexpectedError> {
  return client.query(UsdExchangeRatesQuery, { request });
}

/**
 * Checks if the API has processed a known transaction hash.
 *
 * This is useful to know when cached data has been invalidated after
 * a transaction is complete, as the API uses caching and has an
 * invalidation task that may take 100-200ms longer.
 *
 * ```ts
 * const result = await borrow(client, request)
 *   .andThen(sendWith(wallet))
 *   .andThen(client.waitForTransaction);
 *
 * if (result.isErr()) {
 *   // Handle error
 *   return;
 * }
 *
 * // Check if the transaction has been processed by the API
 * const processed = await hasProcessedKnownTransaction(client, {
 *   txHash: result.value,
 *   operation: OperationType.Borrow
 * });
 *
 * if (processed.isErr()) {
 *   // Handle error
 *   return;
 * }
 *
 * if (processed.value) {
 *   // Transaction processed, cached data is up to date
 *   console.log('Transaction processed, data is fresh');
 * } else {
 *   // Transaction not yet processed, may need to wait
 *   console.log('Transaction not yet processed');
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The request containing transaction hash and operation type to check.
 * @returns True if the transaction has been processed, false otherwise.
 */
export function hasProcessedKnownTransaction(
  client: AaveClient,
  request: HasProcessedKnownTransactionRequest,
): ResultAsync<boolean, UnexpectedError> {
  return client.query(HasProcessedKnownTransactionQuery, { request });
}
