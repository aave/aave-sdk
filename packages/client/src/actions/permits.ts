import type { UnexpectedError } from '@aave/core';
import {
  PermitTypedDataQuery,
  type PermitTypedDataRequest,
  type PermitTypedDataResponse,
} from '@aave/graphql';
import type { ResultAsync } from '@aave/types';
import type { AaveClient } from '../AaveClient';

/**
 * Generates EIP-712 typed data for ERC-20 permit signature.
 *
 * ```ts
 * const result = await permitTypedData(client, {
 *   market: evmAddress('0x87870bca...'),
 *   underlyingToken: evmAddress('0xa0b86a33...'),
 *   amount: '1.5',
 *   chainId: chainId(1),
 *   spender: evmAddress('0x123...'),
 *   owner: evmAddress('0x456...'),
 * });
 *
 * if (result.isOk()) {
 *   const typedData = result.value;
 *   // Use typedData for permit signing
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The permit typed data request parameters.
 * @returns The EIP-712 typed data for permit signature.
 */
export function permitTypedData(
  client: AaveClient,
  request: PermitTypedDataRequest,
): ResultAsync<PermitTypedDataResponse, UnexpectedError> {
  return client.query(PermitTypedDataQuery, { request });
}
