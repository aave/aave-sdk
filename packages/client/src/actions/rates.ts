import {
  type APYSample,
  BorrowAPYHistoryQuery,
  SupplyAPYHistoryQuery,
  type TimeWindow,
} from '@aave/graphql';
import type { ChainId, EvmAddress, ResultAsync } from '@aave/types';
import type { AaveClient } from '../client';
import type { UnexpectedError } from '../errors';

export type BorrowAPYHistoryRequest = {
  /**
   * The markets you want to see based on the chain ids.
   */
  chainId: ChainId;
  /**
   * The user viewing the market (e.g., the connected wallet).
   *
   * If not provided, user fields will not be included.
   */
  underlyingToken: EvmAddress;
  /**
   * The order by clause for the borrow reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  window: TimeWindow;
  /**
   * The order by clause for the supply reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  market: EvmAddress;
};

export function borrowAPRHistory(
  client: AaveClient,
  { chainId, underlyingToken, window, market }: BorrowAPYHistoryRequest,
): ResultAsync<APYSample[] | null, UnexpectedError> {
  return client.query(BorrowAPYHistoryQuery, {
    request: {
      chainId,
      underlyingToken,
      window,
      market,
    },
  });
}

export type SupplyAPYHistoryRequest = {
  /**
   * The markets you want to see based on the chain ids.
   */
  chainId: ChainId;
  /**
   * The user viewing the market (e.g., the connected wallet).
   *
   * If not provided, user fields will not be included.
   */
  underlyingToken: EvmAddress;
  /**
   * The order by clause for the borrow reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  window: TimeWindow;
  /**
   * The order by clause for the supply reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  market: EvmAddress;
};

export function supplyAPRHistory(
  client: AaveClient,
  { chainId, underlyingToken, window, market }: SupplyAPYHistoryRequest,
): ResultAsync<APYSample[] | null, UnexpectedError> {
  return client.query(SupplyAPYHistoryQuery, {
    request: {
      chainId,
      underlyingToken,
      window,
      market,
    },
  });
}
