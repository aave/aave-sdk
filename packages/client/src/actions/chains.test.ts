import { assertOk } from '@aave/client';
import { chains } from '@aave/client/actions';
import { client } from '@aave/client/test-utils';
import { describe, expect, it } from 'vitest';

describe('Given the Aave client', () => {
  describe(`when using the 'chains(client)' action`, () => {
    it('then it should return the supported chains', async () => {
      const result = await chains(client);
      assertOk(result);
      expect(result.value).toMatchInlineSnapshot(`
        [
          {
            "__typename": "Chain",
            "chainId": 1,
            "icon": "https://statics.aave.com/ethereum.svg",
            "name": "Ethereum",
          },
          {
            "__typename": "Chain",
            "chainId": 42161,
            "icon": "https://statics.aave.com/arbitrum.svg",
            "name": "Arbitrum",
          },
          {
            "__typename": "Chain",
            "chainId": 43114,
            "icon": "https://statics.aave.com/avalanche.svg",
            "name": "Avalanche",
          },
          {
            "__typename": "Chain",
            "chainId": 8453,
            "icon": "https://statics.aave.com/base.svg",
            "name": "Base",
          },
          {
            "__typename": "Chain",
            "chainId": 56,
            "icon": "https://statics.aave.com/bnbchain.svg",
            "name": "BSC",
          },
          {
            "__typename": "Chain",
            "chainId": 42220,
            "icon": "https://statics.aave.com/celo.svg",
            "name": "Celo",
          },
          {
            "__typename": "Chain",
            "chainId": 100,
            "icon": "https://statics.aave.com/Gnosis.svg",
            "name": "Gnosis",
          },
          {
            "__typename": "Chain",
            "chainId": 59144,
            "icon": "https://statics.aave.com/linea.svg",
            "name": "Linea",
          },
          {
            "__typename": "Chain",
            "chainId": 1088,
            "icon": "https://statics.aave.com/Metis.svg",
            "name": "Metis",
          },
          {
            "__typename": "Chain",
            "chainId": 10,
            "icon": "https://statics.aave.com/optimism.svg",
            "name": "Optimism",
          },
          {
            "__typename": "Chain",
            "chainId": 137,
            "icon": "https://statics.aave.com/polygon.svg",
            "name": "Polygon",
          },
          {
            "__typename": "Chain",
            "chainId": 534352,
            "icon": "https://statics.aave.com/scroll-network.svg",
            "name": "Scroll",
          },
          {
            "__typename": "Chain",
            "chainId": 1946,
            "icon": "https://statics.aave.com/soneium.svg",
            "name": "Soneium",
          },
          {
            "__typename": "Chain",
            "chainId": 146,
            "icon": "https://statics.aave.com/sonic.svg",
            "name": "Sonic",
          },
          {
            "__typename": "Chain",
            "chainId": 324,
            "icon": "https://statics.aave.com/zksync-era-light.svg",
            "name": "zkSync",
          },
        ]
      `);
    });
  });
});
