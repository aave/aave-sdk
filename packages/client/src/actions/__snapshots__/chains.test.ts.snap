// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the Aave client > When using the 'chains(client)' action > Then it should return the supported chains 1`] = `
[
  {
    "__typename": "Chain",
    "chainId": 1,
    "explorerUrl": "https://etherscan.io",
    "icon": "https://statics.aave.com/ethereum.svg",
    "name": "Ethereum",
  },
  {
    "__typename": "Chain",
    "chainId": 42161,
    "explorerUrl": "https://arbiscan.io",
    "icon": "https://statics.aave.com/arbitrum.svg",
    "name": "Arbitrum",
  },
  {
    "__typename": "Chain",
    "chainId": 43114,
    "explorerUrl": "https://cchain.explorer.avax.network",
    "icon": "https://statics.aave.com/avalanche.svg",
    "name": "Avalanche",
  },
  {
    "__typename": "Chain",
    "chainId": 8453,
    "explorerUrl": "https://basescan.org",
    "icon": "https://statics.aave.com/base.svg",
    "name": "Base",
  },
  {
    "__typename": "Chain",
    "chainId": 56,
    "explorerUrl": "https://bscscan.com",
    "icon": "https://statics.aave.com/bnbchain.svg",
    "name": "BSC",
  },
  {
    "__typename": "Chain",
    "chainId": 42220,
    "explorerUrl": "https://celoscan.io",
    "icon": "https://statics.aave.com/celo.svg",
    "name": "Celo",
  },
  {
    "__typename": "Chain",
    "chainId": 100,
    "explorerUrl": "https://gnosisscan.io",
    "icon": "https://statics.aave.com/Gnosis.svg",
    "name": "Gnosis",
  },
  {
    "__typename": "Chain",
    "chainId": 59144,
    "explorerUrl": "https://lineascan.build",
    "icon": "https://statics.aave.com/linea.svg",
    "name": "Linea",
  },
  {
    "__typename": "Chain",
    "chainId": 1088,
    "explorerUrl": "https://andromeda-explorer.metis.io",
    "icon": "https://statics.aave.com/Metis.svg",
    "name": "Metis",
  },
  {
    "__typename": "Chain",
    "chainId": 10,
    "explorerUrl": "https://optimistic.etherscan.io",
    "icon": "https://statics.aave.com/optimism.svg",
    "name": "Optimism",
  },
  {
    "__typename": "Chain",
    "chainId": 137,
    "explorerUrl": "https://polygonscan.com",
    "icon": "https://statics.aave.com/polygon.svg",
    "name": "Polygon",
  },
  {
    "__typename": "Chain",
    "chainId": 534352,
    "explorerUrl": "https://scrollscan.com",
    "icon": "https://statics.aave.com/scroll-network.svg",
    "name": "Scroll",
  },
  {
    "__typename": "Chain",
    "chainId": 1868,
    "explorerUrl": "https://soneium.blockscout.com",
    "icon": "https://statics.aave.com/soneium.svg",
    "name": "Soneium",
  },
  {
    "__typename": "Chain",
    "chainId": 146,
    "explorerUrl": "https://sonicscan.org",
    "icon": "https://statics.aave.com/sonic.svg",
    "name": "Sonic",
  },
  {
    "__typename": "Chain",
    "chainId": 324,
    "explorerUrl": "https://era.zksync.network",
    "icon": "https://statics.aave.com/zksync-era-light.svg",
    "name": "zkSync",
  },
  {
    "__typename": "Chain",
    "chainId": 5000,
    "explorerUrl": "https://explorer.mantle.xyz",
    "icon": "https://statics.aave.com/mantle.svg",
    "name": "Mantle",
  },
]
`;
